bleachbit/Unix.py

```diff
@@ -39,11 +39,68 @@
 import General
 
 
-class Locales:
+class LocaleCleanerPath:
+    """This represents a path with either a specific folder name or a folder name pattern.
+    It also may contain several compiled regex patterns for localization items (folders or files)
+    and additional LocaleCleanerPaths that get traversed when asked to supply a list of localization
+    items"""
+
+    def __init__(self, location):
+        if location is None:
+            raise RuntimeError("location is none")
+        self.pattern = location
+        self.children = []
+
+    def add_child(self, child):
+        """Adds a child LocaleCleanerPath"""
+        self.children.append(child)
+        return child
+
+    def add_path_filter(self, pre, post):
+        """Adds a filter consisting of a prefix and a postfix
+        (e.g. 'foobar_' and '\.qm' to match 'foobar_en_US.utf-8.qm)"""
+        try:
+            regex = re.compile('^' + pre + Locales.localepattern + post + '$')
+        except Exception as errormsg:
+            raise RuntimeError("Malformed regex '%s' or '%s': %s" % (pre, post, errormsg))
+        self.add_child(regex)
+
+    def get_subpaths(self, basepath):
+        """Returns direct subpaths for this object, i.e. either the named subfolder or all
+        subfolders matching the pattern"""
+        if isinstance(self.pattern, re._pattern_type):
+            return (os.path.join(basepath, p) for p in os.listdir(basepath)
+                    if self.pattern.match(p) and os.path.isdir(os.path.join(basepath, p)))
+        else:
+            path = os.path.join(basepath, self.pattern)
+            return [path] if os.path.isdir(path) else []
+
+    def get_localizations(self, basepath):
+        """Returns all localization items for this object and all descendant objects"""
+        for path in self.get_subpaths(basepath):
+            for child in self.children:
+                if isinstance(child, LocaleCleanerPath):
+                    for res in child.get_localizations(path):
+                        yield res
+                elif isinstance(child, re._pattern_type):
+                    for element in os.listdir(path):
+                        match = child.match(element)
+                        if match is not None:
+                            yield (match.group('locale'), os.path.join(path, element))
+
 
+class Locales:
     """Find languages and localization files"""
 
-    localepattern = r"(?P<locale>[a-z]{2,3})(?:[_-][a-zA-Z]{2,4})?(?:\.[a-zA-Z0-9-]*)?(?:@[a-zA-Z]*)?"
+    # The regular expression to match locale strings and extract the langcode.
+    # See test_locale_regex() in tests/TestUnix.py for examples
+    # This doesn't match all possible valid locale strings to avoid
+    # matching filenames you might want to keep, e.g. the regex
+    # to match jp.eucJP might also match jp.importantfileextension
+    localepattern =\
```

Comment:
I'm not sure how specific the regex should be. it shouldn't match "en.extension", but all possible encodings would be even more complicated.